### ISSUE TITLE
Using libcurl on Linux by pinvokes causes crashes in the libcurl

### ISSUE THREAD

**Issue by janvorli** [dotnet/coreclr#1466]
When using PInvoke to call libcurl functions, random SIGSEGV or malloc reporting double freeing happens in the libcurl. 
Here is a sample application that repros that problem:
```c#
using System;
using System.Net;
using System.Net.Http;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

public class Repro
{
    private const string libcurl = "libcurl";

    [DllImport(libcurl)]
    private static extern int curl_global_init(long flags);

    [DllImport(libcurl)]
    private static extern IntPtr curl_multi_init();

    [DllImport(libcurl)]
    private static extern IntPtr curl_easy_init();

    [DllImport(libcurl)]
    private static extern int curl_multi_add_handle(IntPtr multi_handle, IntPtr easy_handle);

    [DllImport(libcurl, CharSet = CharSet.Ansi)]
    private static extern int curl_easy_setopt(IntPtr curl, int option, string value);

    [DllImport(libcurl)]
    private static extern int curl_multi_perform(IntPtr multi, out int running_handles);

    private const int CURL_GLOBAL_ALL = 3;
    private const int CURLOPT_URL = 10002;

    private static int ThrowIfNotZero(int result, [CallerLineNumber] int line = 0)
    {
        if (result != 0) throw new Exception("Line " + line);
        return result;
    }

    private static IntPtr ThrowIfZero(IntPtr result, [CallerLineNumber] int line = 0)
    {
        if (result == IntPtr.Zero) throw new Exception("Line " + line);
        return result;
    }

    public static void Main()
    {
        ThrowIfNotZero(curl_global_init(CURL_GLOBAL_ALL));
        IntPtr multi = ThrowIfZero(curl_multi_init());

        const int Iters = 100;
        for (int i = 0; i < Iters; i++)
        {
            IntPtr easy = ThrowIfZero(curl_easy_init());
            ThrowIfNotZero(curl_easy_setopt(easy, CURLOPT_URL, "http://www.httpbin.org"));
            ThrowIfNotZero(curl_multi_add_handle(multi, easy));
        }

        Console.WriteLine("Before");
        int running_handles;
        ThrowIfNotZero(curl_multi_perform(multi, out running_handles));
        Console.WriteLine("After");
    }
}
```

A clone of the same code written in C++ works fine:
```C++
#include <stdio.h>
#include <stdlib.h>
#include "curl/curl.h"

int main(int argc, char** argv)
{
       if (curl_global_init(CURL_GLOBAL_ALL) != 0)
             printf("global init failed\n");

       CURLM* multi = curl_multi_init();
       if (multi == NULL)
             printf("couldn't create multi\n");

       int i;
       for (i = 0; i < 100; i++)
       {
             CURL* easy = curl_easy_init();
             if (easy == NULL)
                    printf("couldn't create easy\n");

             if (curl_easy_setopt(easy, CURLOPT_URL, "http://www.httpbin.org") != 0)
                    printf("couldn't set easy\n");

             if (curl_multi_add_handle(multi, easy) != 0)
                    printf("couldn't add easy\n");
       }

       printf("before\n");
       int running;
       if (curl_multi_perform(multi, &running) != 0)
             printf("couldn't perform\n");
       printf("after\n");

       return 0;
}
```

**Comment by janvorli** [comment]
I've finally found the problem cause. I've done a lot of experiments and I've found that I can repro the issue in C++ as well if I load the libcurl using dlopen. While debugging the issue, in one case I've noticed something strange. 
I have seen a place where a pointer was read from memory location, then a function was called with the pointer as a parameter to destroy the object and after returning from the function, the memory location was zeroed. The crash occurred while the code was still running in the called function, I could see that the pointer passed to it was nonzero, but the memory location was already zeroed.
At that point, I have thought it was a memory corruption, since other manifestations of the problem looked like that (corrupted malloc's allocation list or corrupted list of libraries inside the libdl).
I have investigated it further, googling for possible hints and then finally I've found this:
http://libc-help.sourceware.narkive.com/OzUlix5P/segfaults-and-double-free-errors-using-getaddrinfo-a
And really, in many cases the getaddrinfo was on the stack when the crash happened. Here is an excerpt from the web page:
>So what seems to be happening here is the the lock guarding the
critical sections in the _nss_files_gethostbyname4_r function, namely
__libc_lock_lock (lock), is actually no-oping and therefore we're
seeing two threads executing in the same critical section, which
explains why, under observation, the program fails in several
different ways (in the same code path).
This is happening because that lock is in libnss_files.so.2 and this
__libc_lock_lock (lock) translates to a
__libc_maybe_call(__pthread_mutex_lock).
The application has no hard dependency on pthread_mutex_lock. At the
point where this is called the libpthread library hasn't been loaded
and therefore the __pthread_mutex_lock symbol hasn't been resolved
into the PLT and therefore libc_maybe_call simply no-ops the lock.
This would explain why the program works when you
LD_PRELOAD=libpthread.so.2. The symbol is resolved and the PLT entry
is there when the __libc_maybe_call(__pthread_mutex_lock) calls checks
for it.

And now it rang the bell. The issue I've seen with the variable being zeroed was not due to a memory corruption, but rather a locking issue. There was a non-functional lock and another thread ran the function that was destroying the passed in object in parallel with the current thread, which has obviously lead to double freeing.
I’ve tried to add –lpthread option when building my test and the issue went away.
Then I’ve tried to add the pthread library when linking corerun and again, it fixed the issue.
Looking at CMakeLists.txt for corerun / coreconsole, I can see that we already link those to pthread library for FreeBSD, so the fix is just to make that happen unconditionally.
